keras_core/callbacks/backup_and_restore_callback_test.py

```diff
@@ -0,0 +1,224 @@
+    # Checking for invalid backup_dir
+    def test_empty_backup_dir(self):
+        with self.assertRaisesRegex(
+            ValueError, expected_regex="Empty " "`backup_dir`"
+        ):
+            callbacks.BackupAndRestore(file_path=None)
+
+    # Checking save_freq and save_before_preemption both unset
+    def test_save_set_error(self):
+        with self.assertRaisesRegex(
+            ValueError,
+            expected_regex="`save_freq` or "
+            "`save_before_preemption` "
+            ""
+            "must be set",
+        ):
+            callbacks.BackupAndRestore(
+                file_path="backup_dir",
+                save_freq=None,
+                save_before_preemption=False,
+            )
+
+    # Check invalid save_freq, both string and non integer
+    def test_save_freq_unknown_error(self):
+        with self.assertRaisesRegex(
+            ValueError, expected_regex="Unrecognized save_freq"
+        ):
+            callbacks.BackupAndRestore(
+                file_path="backup_dir", save_freq="batch"
+            )
+
+        with self.assertRaisesRegex(
+            ValueError, expected_regex="Unrecognized save_freq"
+        ):
+            callbacks.BackupAndRestore(file_path="backup_dir", save_freq=0.15)
+
+    # Checking if after interruption, correct model params and
+    # weights are loaded in step-wise backup
+    @pytest.mark.requires_trainable_backend
+    def test_best_case_step(self):
+        def make_model():
+            np.random.seed(1337)
+            model = Sequential(
+                [
+                    layers.Dense(2, activation="relu"),
+                    layers.Dense(1),
+                ]
+            )
+            model.compile(
+                loss="mse",
+                optimizer="sgd",
+                metrics=["mse"],
+            )
+            return model
+
+        temp_dir = self.get_temp_dir()
+        filepath = os.path.join(temp_dir, "subdir", "checkpoint.weights.h5")
+        file_utils.rmtree(filepath)
+        self.assertFalse(os.path.exists(filepath))
+
+        model = make_model()
+        cbk = callbacks.BackupAndRestore(file_path=filepath, save_freq=1)
+
+        x_train = np.random.random((10, 3))
+        y_train = np.random.random((10, 1))
+
+        try:
+            model.fit(
+                x_train,
+                y_train,
+                batch_size=4,
+                callbacks=[
+                    cbk,
+                    InterruptingCallback(steps_int=2, epoch_int=None),
+                ],
+                epochs=2,
+                verbose=0,
+            )
+        except RuntimeError:
+            self.assertTrue(os.path.exists(filepath))
+            self.assertEqual(cbk._current_epoch, 0)
+            self.assertEqual(cbk._last_batch_seen, 1)
+
+            hist = model.fit(
+                x_train, y_train, batch_size=4, callbacks=[cbk], epochs=5
+            )
+
+            self.assertEqual(cbk._current_epoch, 4)
```

Comment:
Can you also assert that `len(hist["loss"]) == 3` to confirm that it only ran for 3 additional epochs and not the entire 5 epochs again?  I would expect that assertion to fail since we don't seem to restore the `epoch` in `on_train_begin`.

Reply:
@fchollet - What do you think we should do?  Is there a good way we can set the `iniital_epoch` of the model?  Also, we are not restoring the Optimizer state that `tf.keras` had restored when it saved it in training checkpoint.  Any suggestions.

Reply:
Checked with Francois and addressed both questions - 
1. Model saving also includes optimizer, so optimizer state is also restored when backup is restored.
2. On setting `initial_epoch`, we may need to add that hook on `model` before we can set it in the `on_train_begin` of `BackupAndRestore`.  We will address that in a subsequent PR.
We can make this conversation resolved.